File: leet-code/videoStitching.js
/**
 * @param {number[][]} clips
 * @param {number} T
 * @return {number}
 */
var videoStitching = function (clips, T) {
  let ans = 0
  const maxRange = new Array(T).fill(0)

  // 记录每个开始位置的最大结束位置
  for (const item of clips) {
    const [l, r] = item
    if (r > maxRange[l]) {
      maxRange[l] = r
    }
  }

  let pre = 0,
    last = 0
  for (let i = 0; i < maxRange.length; i++) {
    const val = maxRange[i]
    if (val > last) {
      last = val
    }

    if (i === last) {
      return -1
    }

    if (i === pre) {
      ans++
      pre = last
    }
  }
  return ans
}

console.log(
  videoStitching(
    [
      [0, 2],
      [4, 6],
      [8, 10],
      [1, 9],
      [1, 5],
      [5, 9]
    ],
    10
  )
)

File: leet-code/videoStitching.go
package main

import "math"

func videoStitching(clips [][]int, T int) int {
	ans := 0
	maxRange := make([]int, T)
	for _, val := range clips {
		l, r := val[0], val[1]
		if l < T && r > maxRange[l] {
			maxRange[l] = r
		}
	}

	pre := 0
	last := 0
	for i := range maxRange {
		if last < maxRange[i] {
			last = maxRange[i]
		}
		if i == last {

			return -1
		}
		if i == pre {
			ans++
			pre = last
		}
	}
	return ans
}

func videoStitching2(clips [][]int, T int) int {
	dp := make([]int, T+1)
	for i := range dp {
		dp[i] = math.MaxInt64 - 1
	}
	dp[0] = 0

	for i := 1; i <= T; i++ {
		for _, v := range clips {
			l, r := v[0], v[1]
			if l < i && i <= r && dp[l]+1 < dp[i] {
				dp[i] = dp[l] + 1
			}
		}
	}

	if dp[T] == math.MaxInt64-1 {
		return -1
	}
	return dp[T]
}
